File: Java/Endpoint Examples/Multipart Payload/RequestStatus.java
import io.github.cdimascio.dotenv.Dotenv;
import java.io.File;
import java.io.IOException;
import java.util.concurrent.TimeUnit;
import okhttp3.*;
import org.json.JSONObject;

public class RequestStatus {

  // By default, we use the US-based API service. This is the primary endpoint for global use.
  private static final String API_URL = "https://api.pdfrest.com";

  // For GDPR compliance and enhanced performance for European users, you can switch to the EU-based
  // service by uncommenting the URL below.
  // For more information visit https://pdfrest.com/pricing#how-do-eu-gdpr-api-calls-work
  // private static final String API_URL = "https://eu-api.pdfrest.com";

  // Specify your API key here, or in the environment variable PDFREST_API_KEY.
  // You can also put the environment variable in a .env file.
  private static final String DEFAULT_API_KEY = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx";
  // Specify the path to your file here, or as the first argument when running the program.
  private static final String DEFAULT_FILE_PATH = "/path/to/file.pdf";

  public static void main(String[] args) {

    File inputFile;
    if (args.length > 0) {
      inputFile = new File(args[0]);
    } else {
      inputFile = new File(DEFAULT_FILE_PATH);
    }

    final Dotenv dotenv = Dotenv.configure().ignoreIfMalformed().ignoreIfMissing().load();
    String apiKey = dotenv.get("PDFREST_API_KEY", DEFAULT_API_KEY);

    // Using PDF/A as an arbitrary example, send a request with a 'response-type' header.
    String pdfaResponse = getPdfaResponse(inputFile, apiKey);
    JSONObject pdfaJson = new JSONObject(pdfaResponse);
    if (pdfaJson.has("error")) {
      System.out.println("Error during PDFA call: " + pdfaJson.getString("error"));
    } else {
      // Get a request ID from the response.
      String requestID = pdfaJson.getString("requestId");

      // Check the request status.
      String requestStatusResponse = getRequestStatusResponse(requestID, apiKey);
      JSONObject requestStatusJson = new JSONObject(requestStatusResponse);

      // If still pending, check periodically for updates.
      while (requestStatusJson.getString("status").equals("pending")) {
        final int delay = 5000;
        try {
          Thread.sleep(delay);
          requestStatusResponse = getRequestStatusResponse(requestID, apiKey);
          requestStatusJson = new JSONObject(requestStatusResponse);
        } catch (InterruptedException e) {
          System.out.println(e);
        }
      }
    }
  }

  private static String getPdfaResponse(File inputFile, String apiKey) {

    final RequestBody inputFileRequestBody =
        RequestBody.create(inputFile, MediaType.parse("application/pdf"));
    RequestBody requestBody =
        new MultipartBody.Builder()
            .setType(MultipartBody.FORM)
            .addFormDataPart("file", inputFile.getName(), inputFileRequestBody)
            .addFormDataPart("output_type", "PDF/A-2u")
            .addFormDataPart("output", "pdfrest_pdfa")
            .build();
    Request request =
        new Request.Builder()
            .header("Api-Key", apiKey)
            .header("response-type", "requestId")
            .url(API_URL + "/pdfa")
            .post(requestBody)
            .build();
    try {
      OkHttpClient client =
          new OkHttpClient().newBuilder().readTimeout(60, TimeUnit.SECONDS).build();

      Response response = client.newCall(request).execute();
      System.out.println("Result code " + response.code());
      String responseBody = response.body().string();
      System.out.println(prettyJson(responseBody));
      return responseBody;
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  private static String getRequestStatusResponse(String requestId, String apiKey) {
    String urlString = String.format(API_URL + "/request-status/%s", requestId);
    Request request = new Request.Builder().header("Api-Key", apiKey).url(urlString).get().build();
    try {
      OkHttpClient client =
          new OkHttpClient().newBuilder().readTimeout(60, TimeUnit.SECONDS).build();
      Response response = client.newCall(request).execute();
      System.out.println("Result code " + response.code());
      String responseBody = response.body().string();
      System.out.println(prettyJson(responseBody));
      return responseBody;
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  private static String prettyJson(String json) {
    // https://stackoverflow.com/a/9583835/11996393
    return new JSONObject(json).toString(4);
  }
}

File: Python/Endpoint Examples/Multipart Payload/request-status.py
from requests_toolbelt import MultipartEncoder
import requests
import json
import time

# By default, we use the US-based API service. This is the primary endpoint for global use.
api_url = "https://api.pdfrest.com"

# For GDPR compliance and enhanced performance for European users, you can switch to the EU-based service by uncommenting the URL below.
# For more information visit https://pdfrest.com/pricing#how-do-eu-gdpr-api-calls-work
#api_url = "https://eu-api.pdfrest.com"

api_key = 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx' # place your api key here

pdfa_endpoint_url = api_url+'/pdfa'

mp_encoder_pdfa = MultipartEncoder(
    fields={
        'file': ('file_name.pdf', open('/path/to/file.pdf', 'rb'), 'application/pdf'),
        'output_type': 'PDF/A-1b',
    }
)

# Send a request to a pdfRest tool with the Response-Type header to get a request ID,
pdfa_headers = {
    'Accept': 'application/json',
    'Content-Type': mp_encoder_pdfa.content_type,
    'Response-Type': "requestId",
    'Api-Key': api_key
}

print("Sending POST request to pdfa endpoint...")
response = requests.post(pdfa_endpoint_url, data=mp_encoder_pdfa, headers=pdfa_headers)

print("Response status code: " + str(response.status_code))

if response.ok:

    response_json = response.json()
    request_id = response_json["requestId"]
    api_polling_endpoint_url = f'{api_url}/request-status/{request_id}'

    headers = {
        'Api-Key': api_key
    }

    print("Sending GET request to request-status endpoint...")
    response = requests.get(api_polling_endpoint_url, headers=headers)

    print("Response status code: " + str(response.status_code))

    if response.ok:
        response_json = response.json()
        while response_json["status"] == "pending":
            # This example will get the request status every 5 seconds until the request is completed.
            print(json.dumps(response_json, indent = 2))
            time.sleep(5)
            response = requests.get(api_polling_endpoint_url, headers=headers)
            response_json = response.json()
        print(json.dumps(response_json, indent = 2))
    else:
        print(response.text)
